src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -0,0 +1,823 @@
+@end
+
+@implementation MTRDevice
+
+- (instancetype)initWithDeviceID:(uint64_t)deviceID
+                deviceController:(MTRDeviceController *)deviceController
+                           queue:(dispatch_queue_t)queue
+{
+    if (self = [super init]) {
+        _lock = OS_UNFAIR_LOCK_INIT;
+        _deviceID = deviceID;
+        _deviceController = deviceController;
+        _queue = queue;
+        _readCache = [NSMutableDictionary dictionary];
+        _expectedValueCache = [NSMutableDictionary dictionary];
+        _asyncCallbackWorkQueue = [[MTRAsyncCallbackWorkQueue alloc] initWithContext:self queue:queue];
+    }
+    return self;
+}
+
++ (instancetype)deviceWithDeviceID:(uint64_t)deviceID deviceController:(MTRDeviceController *)deviceController
+{
+    return [deviceController deviceForDeviceID:deviceID];
+}
+
+- (void)connectAndPerformAsync:(MTRDevicePerformAsyncBlock)asyncBlock
+{
+    [_deviceController getBaseDevice:_deviceID
+                               queue:_queue
+                   completionHandler:^(MTRBaseDevice * _Nullable device, NSError * _Nullable error) {
+                       asyncBlock(device);
+                   }];
+}
+
+#pragma mark Subscription
+- (void)performBlockForDelegates:(void (^)(id<MTRDeviceSubscriptionDelegate>))block
+{
+    os_unfair_lock_lock(&self->_lock);
+    NSSet * delegatesToCall = [_subscribers copy];
+    os_unfair_lock_unlock(&self->_lock);
+
+    NSMutableSet * lostDelegates = [NSMutableSet set];
+    for (MTRWeakReference<id<MTRDeviceSubscriptionDelegate>> * delegateWeakReference in delegatesToCall) {
+        id<MTRDeviceSubscriptionDelegate> delegate = delegateWeakReference.strongObject;
+        if (!delegate) {
+            [lostDelegates addObject:delegateWeakReference];
+        }
+
+        block(delegate);
+    }
+
+    os_unfair_lock_lock(&self->_lock);
+    [_subscribers minusSet:lostDelegates];
+    os_unfair_lock_unlock(&self->_lock);
+
+    // Options to consider when no more subscribers are alive:
+    //   A) tear down subscription
+    //   B) keep subscription until error, and wait for client to resubscribe
+    //   C) keep subscription / resubscribe on error, until EOL MTRDevice
+}
+
+- (void)_handleSubscriptionEstablished
+{
+    [self performBlockForDelegates:^(id<MTRDeviceSubscriptionDelegate> delegate) {
+        [delegate subscriptionEstablished];
+    }];
+}
+
+- (void)_handleSubscriptionError:(NSError *)error
+{
+    _subscriptionActive = NO;
+
+    [self performBlockForDelegates:^(id<MTRDeviceSubscriptionDelegate> delegate) {
+        [delegate subscriptionEndedWithError:error];
+    }];
+    [_subscribers removeAllObjects];
+}
+
+- (void)_reportAttributes:(NSArray<NSDictionary<NSString *, id> *> *)attributes
+{
+    if (attributes.count) {
+        [self performBlockForDelegates:^(id<MTRDeviceSubscriptionDelegate> delegate) {
+            [delegate subscriptionReceivedAttributeReport:attributes];
```

Comment:
Do we need to worry about the fact that a delegate could mutate the data structure, or is it all immutable in the end because NSNumber/NSString/NSData are immutable?